floss/language/go/extract.py

```diff
@@ -0,0 +1,699 @@
+            | \x8D \x1D  # 8d 1d aa aa 00 00       lea    ebx,ds:0xaaaa
+            | \x8D \x0D  # 8d 0d aa aa 00 00       lea    ecx,ds:0xaaaa
+            | \x8D \x15  # 8d 15 aa aa 00 00       lea    edx,ds:0xaaaa
+            | \x8D \x35  # 8d 35 aa aa 00 00       lea    esi,ds:0xaaaa
+            | \x8D \x3D  # 8d 3d aa aa 00 00       lea    edi,ds:0xaaaa
+        )
+        (?P<address>....)
+        """,
+        re.DOTALL + re.VERBOSE,
+    )
+
+    for match in absolute_insn_re.finditer(buf):
+        address_bytes = match.group("address")
+        address = struct.unpack("<I", address_bytes)[0]
+
+        yield address
+
+
+def find_lea_xrefs(pe: pefile.PE) -> Iterable[VA]:
+    """
+    scan the executable sections of the given PE file
+    for LEA instructions that reference valid memory addresses,
+    yielding the virtual addresses.
+    """
+    low, high = get_image_range(pe)
+
+    for section in pe.sections:
+        if not section.IMAGE_SCN_MEM_EXECUTE:
+            continue
+
+        code = section.get_data()
+
+        if pe.FILE_HEADER.Machine == pefile.MACHINE_TYPE["IMAGE_FILE_MACHINE_AMD64"]:
+            xrefs = find_amd64_lea_xrefs(code, section.VirtualAddress + pe.OPTIONAL_HEADER.ImageBase)
+        elif pe.FILE_HEADER.Machine == pefile.MACHINE_TYPE["IMAGE_FILE_MACHINE_I386"]:
+            xrefs = find_i386_lea_xrefs(code)
+        else:
+            raise ValueError("unhandled architecture")
+
+        for xref in xrefs:
+            if low <= xref < high:
+                yield xref
+
+
+@dataclass(frozen=True)
+class StructString:
+    """
+    a struct String instance.
+
+
+    ```go
+        // String is the runtime representation of a string.
+        // It cannot be used safely or portably and its representation may
+        // change in a later release.
+        //
+        // Unlike reflect.StringHeader, its Data field is sufficient to guarantee the
+        // data it references will not be garbage collected.
+        type String struct {
+            Data unsafe.Pointer
+            Len  int
+        }
+    ```
+
+    https://github.com/golang/go/blob/36ea4f9680f8296f1c7d0cf7dbb1b3a9d572754a/src/internal/unsafeheader/unsafeheader.go#L28-L37
+    """
+
+    address: VA
+    length: int
+
+
+def get_max_section_size(pe: pefile.PE) -> int:
+    """get the size of the largest section, as seen on disk."""
+    return max(map(lambda s: s.SizeOfRawData, pe.sections))
+
+
+def get_struct_string_candidates_with_pointer_size(pe: pefile.PE, buf: bytes, psize: int) -> Iterable[StructString]:
+    """
+    scan through the given bytes looking for pairs of machine words (address, length)
+    that might potentially be struct String instances.
+
+    we do some initial validation, like checking that the address is valid
+    and the length is reasonable; however, we don't validate the encoded string data.
+    """
+    if psize == 32:
+        format = "I"
```

Comment:
since this is for an array (not struct) the formats are declared here: https://docs.python.org/3/library/array.html

I is for 2 byte words and L is for 4 bytes (and Q for 8 bytes).

Reply:
and it seems like we want "L" but I used "I" which seems like a mistake. i should have tested this. 

when you say it doesn't work, what are you observing?